src/oc/hoc_oop.c

```diff
@@ -501,18 +501,79 @@ Object** hoc_temp_objvar(Symbol* symtemp, void* v){
-  unreffed. Not going to work if constructors themselves create new objects
-  before the error. Or if the destructor doesn't work with a partially filled
-  object.
+  unreffed. To handle the case of constructors themselves creating new objects
+  before the error or intervening recovery of error by a callee recovering from
+  execerror, the incomplete new object is put on a stack along with the
+  current longjump target, and removed from the stack when the object
+  is complete. There could be a problem if the destructor doesn't work
+  with a partially constructed object. In case of a execerror before newobj1
+  completion, all partially constructed objects with a longjump handle equal
+  to the current longjump handle are unreffed.
 **/
 
-static Object* hoc_newobj1_err_;
-void hoc_newobj1_err() { /* called from hoc_execerror */
-	if (hoc_newobj1_err_) {
-		Object* ob = hoc_newobj1_err_;
-		hoc_newobj1_err_ = NULL;
-		hoc_obj_unref(ob);
-	}
+#define NEWOBJ1_ERR_SIZE 32 /* starts with this size, and doubles on overflow */
+typedef struct {
+  Object* ob;
+  void* oji;
+} newobj1_err_t;
+	
+extern void* nrn_get_oji();
+extern void* nrn_get_hoc_jmp();
+extern void (*oc_jump_target_)();
+static int newobj1_err_index_; /* stack index */
+static int newobj1_err_size_;
+static newobj1_err_t* newobj1_err_; /* stack of newobj1_err_t */
+
+/** save partially constructed object and controlling longjump handle **/
+static void push_newobj1_err(Object* ob) {
+  if (newobj1_err_index_ >= newobj1_err_size_) {
+    if (newobj1_err_size_ == 0) {
+      newobj1_err_size_ = NEWOBJ1_ERR_SIZE;
+      newobj1_err_ = (newobj1_err_t*)calloc(newobj1_err_size_, sizeof(newobj1_err_t));
+      assert(newobj1_err_);
+    }else{
+      newobj1_err_size_ *= 2;
+      newobj1_err_ = (newobj1_err_t*)realloc(newobj1_err_, newobj1_err_size_*sizeof(newobj1_err_t));
+      assert(newobj1_err_);
+    }
+  }
+
+  newobj1_err_t* ne = newobj1_err_ + newobj1_err_index_++;
+  ne->ob = ob;
+  ne->oji = oc_jump_target_ ? nrn_get_oji() : nrn_get_hoc_jmp();
+}
+
+/** pop the now fully constructed object **/
+void pop_newobj1_err() {
+  --newobj1_err_index_;
+  assert(newobj1_err_index_ >= 0);
+}
+
+/** unref partially constructed objects controlled by current longjump handle **/
+void hoc_newobj1_err(void* jmp) { /* called from hoc_execerror */
+  if (newobj1_err_index_ > 0) {
+    int i;
+    /* Note: for the case of pure hoc, there may not be an oc_jump_target_
+       in which case jmp will get set to the hoc.c controlling jmp_buf
+    */
+    void* oji = (jmp == (void*)oc_jump_target_) ? nrn_get_oji() : nrn_get_hoc_jmp();
+#if 0 /* useful for debugging */
+    printf("newobj1_err    jmp=%p  current oji=%p\n", jmp, oji);
+    for (i = 0; i < newobj1_err_index_; ++i) {
+      newobj1_err_t* ne = newobj1_err_ + i;
+      printf("newobj1_err[%d] with oji=%p ob=%p %s\n", i, ne->oji, ne->ob, hoc_object_name(ne->ob));
+    }
+#endif
```

Comment:
Should this be kept here permanently i.e. is this often useful?

Reply:
I'm removing it